src/PowerShellEditorServices/Session/PromptNest.cs

```diff
@@ -0,0 +1,564 @@
+        /// </param>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>The <see cref="RunspaceHandle" /> for the current frame.</returns>
+        internal RunspaceHandle GetRunspaceHandle(CancellationToken cancellationToken, bool isReadLine)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Also grab the main runspace handle if this is for a ReadLine pipeline and the runspace
+            // is in process.
+            if (isReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                GetRunspaceHandleImpl(cancellationToken, isReadLine: false);
+            }
+
+            return GetRunspaceHandleImpl(cancellationToken, isReadLine);
+        }
+
+
+        /// <summary>
+        /// Get the <see cref="RunspaceHandle" /> for the current <see cref="PromptNestFrame" />.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken" /> that will be checked prior to
+        /// completing the returned task.
+        /// </param>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>
+        /// A <see cref="Task{RunspaceHandle}" /> object representing the asynchronous operation.
+        /// The <see cref="Task{RunspaceHandle}.Result" /> property will return the
+        /// <see cref="RunspaceHandle" /> for the current frame.
+        /// </returns>
+        internal async Task<RunspaceHandle> GetRunspaceHandleAsync(CancellationToken cancellationToken, bool isReadLine)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Also grab the main runspace handle if this is for a ReadLine pipeline and the runspace
+            // is in process.
+            if (isReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                await GetRunspaceHandleImplAsync(cancellationToken, isReadLine: false);
+            }
+
+            return await GetRunspaceHandleImplAsync(cancellationToken, isReadLine);
+        }
+
+        /// <summary>
+        /// Releases control of the runspace aquired via the <see cref="RunspaceHandle" />.
+        /// </summary>
+        /// <param name="runspaceHandle">
+        /// The <see cref="RunspaceHandle" /> representing the control to release.
+        /// </param>
+        internal void ReleaseRunspaceHandle(RunspaceHandle runspaceHandle)
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            ReleaseRunspaceHandleImpl(runspaceHandle.IsReadLine);
+            if (runspaceHandle.IsReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                ReleaseRunspaceHandleImpl(isReadLine: false);
```

Comment:
Why do you call this function twice?

Reply:
Both the main thread handle and the ReadLine handle need to be released.